src/ir/ir_traversing_visitor.h

```diff
@@ -24,47 +25,74 @@ namespace raksha::ir {
 // A visitor that also traverses the children of a node and allows performing
 // different actions before (PreVisit) and after (PostVisit) the children are
 // visited. Override any of the `PreVisit` and `PostVisit` methods as needed.
-template <typename Derived>
-class IRTraversingVisitor : public IRVisitor<Derived> {
+template <typename Derived, typename Result = Unit>
+class IRTraversingVisitor : public IRVisitor<Derived, Result> {
  public:
   virtual ~IRTraversingVisitor() {}
 
   // Invoked before all the children of `module` is visited.
-  virtual void PreVisit(const Module& module) {}
+  virtual Result PreVisit(const Module& module) {
+    return Result();
+  }
   // Invoked after all the children of `module` is visited.
-  virtual void PostVisit(const Module& module) {}
+  virtual Result PostVisit(const Module& module, Result result) {
```

Comment:
Somehow make it clear that `result` is from after the in-order traversal? e.g.,
```suggestion
  virtual Result PostVisit(const Module& module, Result in_order_result) {
```

Reply:
There were some other instances to change so I've made those renames myself, thanks for the diff though